spec/terrafying/components/vpc_spec.rb

```diff
@@ -0,0 +1,217 @@
+    expect(ssh_security_group[:egress].count).to eq(1)
+
+    rule = {
+      from_port: 22,
+      to_port: 22,
+      protocol: "tcp",
+      cidr_blocks: [cidr],
+    }
+
+    expect(ssh_security_group[:ingress][0]).to eq(rule)
+    expect(ssh_security_group[:egress][0]).to eq(rule)
+  end
+
+  context "peer_with" do
+
+    it "should raise an error if the cidrs are overlapping" do
+      vpc_a = Terrafying::Components::VPC.create("a", "10.0.0.0/16")
+      vpc_b = Terrafying::Components::VPC.create("b", "10.0.0.0/20")
+
+      expect {
+        vpc_a.peer_with(vpc_b)
+      }.to raise_error(RuntimeError)
+    end
+
+    it "should create routes in both VPCs" do
+      vpc_a = Terrafying::Components::VPC.create("a", "10.0.0.0/16")
+      vpc_b = Terrafying::Components::VPC.create("b", "10.1.0.0/16")
+
+      original_route_count = vpc_a.output_with_children["resource"]["aws_route"].count
+
+      vpc_a.peer_with(vpc_b)
+
+      num_new_routes = vpc_a.output_with_children["resource"]["aws_route"].count - original_route_count
+
+      expect(num_new_routes).to eq(vpc_a.subnets.count * vpc_a.azs.count + vpc_b.subnets.count * vpc_b.azs.count)
+    end
+
+    it "should allow users to limit the subnets that are peered" do
+      vpc_a = Terrafying::Components::VPC.create("a", "10.0.0.0/16")
+      vpc_b = Terrafying::Components::VPC.create("b", "10.1.0.0/16")
+
+      original_route_count = vpc_a.output_with_children["resource"]["aws_route"].count
+
+      vpc_a.peer_with(vpc_b, { our_subnets: [vpc_a.subnets[:public]],
+                               their_subnets: [vpc_b.subnets[:public]] })
+
+      num_new_routes = vpc_a.output_with_children["resource"]["aws_route"].count - original_route_count
+
+      expect(num_new_routes).to eq(1 * vpc_a.azs.count + 1 * vpc_b.azs.count)
+    end
+
+  end
+
+  context "extract_subnet!" do
+
+    it "should limit the size of a subnet to a /28" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      cidr = vpc.extract_subnet!(30)
+
+      expect(cidr).to match(/[\.0-9]+\/28/)
+    end
+
+    it "should raise when there are no subnets left" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+
+      249.times {
+        vpc.extract_subnet!(24)
+      }
+
+      expect {
+        vpc.extract_subnet!(24)
+      }.to raise_error(RuntimeError)
+    end
+
+  end
+
+  context "allocate_subnet!" do
+
+    it "should create subnets for each availability zone" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      subnets = vpc.allocate_subnets!("asdf")
+
+      expect(subnets.count).to eq(vpc.azs.count)
+    end
+
+    it "should attach an internet gateway if subnet is public" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      subnets = vpc.allocate_subnets!("asdf", { public: true })
+
+      output = vpc.output_with_children
+
+      route = output["resource"]["aws_route"].values.select { |route|
```

Comment:
Ruby trick: `select {...}.first` is just `detect`. Doesn't need to be changed, but good to know.

Reply:
I will bear this in mind. Thanks for the pointer!